### ISSUE TITLE
magazzino e giacenze

### ISSUE THREAD

**Issue by madbob** [madbob/GASdottoNG#131]
Derivazione di https://github.com/madbob/GASdotto/issues/64

- permettere di marcare un fornitore con un flag "gestione giacenze"
- se attivato, mostrare un parametro "quantità disponibile" nei prodotti
- aggiungere un nuovo pannello "Magazzino" che mostri l'elenco dei fornitori attivati
- internamente, mantenere un ordine perennemente aperto con la possibilita' di prenotare (e consegnare anche se l'ordine non e' chiuso)
- le prenotazioni fatte a tale ordine permanente incidono sulla "quantita' disponibile"
- per gli ordini regolari fatti al fornitore, aggiungere per i referenti la possibilita' di creare una prenotazione "speciale" che, quando consegnata, vada ad incrementare la "quantita' disponibile" dei prodotti

**Comment by riccardi1997** [comment]
Ciao @madbob sto lavorando a questa issue, ma il pannello "magazzino" dove vorresti aggiungerlo? Inoltre, solo per l'amministratore del gas suppongo. Potresti darmi maggiori info a riguardo?

**Comment by madbob** [comment]
Bella domanda: piu' o meno ho pensato ad un workflow (quello descritto sopra) ma non ai dettagli.

Il pannello "Magazzino" potrebbe essere nel menu principale, in modo da distinguerlo dagli ordini regolari, e ci dovrebbero finire gli ordini "virtuali" (altrimenti si mischiano con quelli "regolari", e potrebbe essere un po' incasinato) accessibili secondo gli stessi criteri con cui si accede al pannello "Ordini" (ruoli e fornitori assegnati).

All'atto pratico un ordine "virtuale" dovrebbe essere simile ad un ordine "regolare", con eccezioni notevoli:
- non ci sono date di apertura, chiusura e consegna
- non c'e' uno stato aperto/chiuso
- probabilmente non sono applicabili i modificatori
- le consegne sono sempre accessibili e consegnabili man mano che arrivano

Insomma: c'e' molto da inventare :-P

Ti interessa questa funzione particolare? Nel tuo GAS si usa questo sistema? Forse val la pena prima capire come funziona il processo, e poi implementarlo.

**Comment by riccardi1997** [comment]
No, stavo solo cercando un punto da cui partire per essere utile al tuo progetto :-P
Io non so se hai avuto modo, ma immagino di sì, di confrontarti con i vari GAS e di capire quando avviene un ordine classico e quando diversamente un ordine virtuale. Credo che tu abbia già un'idea ben precisa, riusciresti a darmi delucidazioni in merito?

**NOTA:** Per la similarità che c'è tra i due diversi tipi di ordine possiamo anche pensare di far implementare alle classi ordini un'interfaccia, ma prima pensiamo a far funzionare il tutto.

**Comment by madbob** [comment]
Beh, se vuoi far qualcosa di utile per il gusto di farlo, temo tu abbia scelto la issue sbagliata :-P
Questa e' decisamente complicata, e - come avrai intuito - io stesso ho difficolta' a farne una analisi. E' qui in sospeso da tempo proprio perche' e' una funzione che piu' di uno ha menzionato in passato, ma non ho mai avuto l'opportunita' di confrontarmi realmente con un GAS che la userebbe.

Se vuoi smanettare un poco, ti posso consigliare di iniziare ad esempio da:
- https://github.com/madbob/GASdottoNG/issues/123 - aggiungerei un flag nel pannello di esportazione del "Dettaglio Consegne" da gestire nell'apposita funzione di generazione di PDF e CSV (dovrebbe essere sufficiente prelevare le prenotazioni "top level" oppure tutte quante indiscriminatamente)
- https://github.com/madbob/GASdottoNG/issues/132 - c'e' gia' un punto in cui vengono fatti i controlli multi-GAS, da integrare con gli altri parametri (inclusa la quantita' minima, introdotta relativamente da poco e non citata nella issue)
- https://github.com/madbob/GASdottoNG/issues/187 - questo sembra semplice ma non e' proprio banalissimo, in quanto vanno riapplicati i modificatori ad ogni interrogazione dei dati delle statistiche
